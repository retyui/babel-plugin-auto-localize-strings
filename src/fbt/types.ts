export interface FbtOptions {
  objectPropertyAsGetter?: boolean;

  jsxFormat?: "CamelCase" | " xmlNamespaces";

  // fbt(...)
  description?: string;
  // author?: string;
  // preserveWhitespace: boolean;
  // subject: string;
  // common?: boolean;
  // doNotExtract?: boolean;
}
