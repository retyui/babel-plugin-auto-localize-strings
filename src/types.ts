import type { FbtOptions } from "./fbt/types";

export interface Options {
  mode?: "fbt" | "i18next";
  fbtOptions?: FbtOptions;
}

export interface State {
  opts: Options;
}

export interface ThisScope {
  importInjected: boolean;
}
