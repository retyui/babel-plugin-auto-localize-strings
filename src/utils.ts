import prettier from "prettier";
import { transformSync } from "@babel/core";
import plugin from "./index";
import type { Options } from "./types";

export function transform(source: string, options: Options = {}): string {
  const { code } = transformSync(source, {
    filename: "file.tsx",
    ast: false,
    plugins: [[plugin, options]],
    sourceType: "module",
    babelrc: false,
    configFile: false,
    compact: false,
  })!;

  return prettier
    .format(code!, { parser: "babel" })
    .replace(`"use strict";\n`, "");
}
