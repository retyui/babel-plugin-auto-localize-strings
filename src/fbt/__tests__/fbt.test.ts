import { transform } from "../../utils";
import { FbtOptions } from "../types";

function runTest({ source, ...options }: { source: string } & FbtOptions) {
  expect(
    transform(source, { mode: "fbt", fbtOptions: options })
  ).toMatchSnapshot();
}

it("object property", () => {
  runTest({
    source: `
const test = {prop: 'test'};
  `,
  });
});

it("literal variable", () => {
  runTest({
    source: `
const test = 'text';
  `,
  });
});

it("literal variable with custom description", () => {
  runTest({
    description: "my desc",
    source: `
const test = 'text';
  `,
  });
});

it("literal variable if needed", () => {
  runTest({
    source: `
const test = 'text';
const test2 = fbt('text2', 'my description');
  `,
  });
});

it("prevent transform fbt options", () => {
  runTest({
    source: `
const test2 = fbt('text2', 'my description', {project:"foo"});
  `,
  });
});

it("jsx attributes", () => {
  runTest({
    source: `
<input title="need to translate" className="ignore"/>
  `,
  });
});

it("jsx simple text", () => {
  runTest({
    source: `
<span>Text</span>;
<span>{'Text'}</span>;
<span>{"Text"}</span>;
<span>{\`Text\`}</span>;
  `,
  });
});

it("jsx with custom description", () => {
  runTest({
    description: "my desc",
    source: `
<span>Text</span>
  `,
  });
});

it("ignore empty jsx text", () => {
  runTest({
    source: `
<div>
  <div/>
</div>
  `,
  });
});

it("auto parametrization on end", () => {
  runTest({
    source: `
<span>text <a>end</a></span>
  `,
  });
});

it("auto parametrization on start", () => {
  runTest({
    source: `
<span><a>start</a> text</span>
  `,
  });
});

it("add jsx wrapper on same deep level", () => {
  runTest({
    source: `
<>
  <div>
    <div>
      <span>
        <a>1 link</a> from friends 1
      </span>    
    </div>
  </div>
  <div>
    <div>
      <span>
        2 from friends <a>2 link</a>
      </span>    
    </div>
  </div>
</>
  `,
  });
});

it("2 levels of nested jsx elements", () => {
  runTest({
    source: `
<div>
  text
  <level1>
    <level2>222</level2>
  </level1>
</div>;
  `,
  });
});

it("3 levels of nested jsx elements", () => {
  runTest({
    source: `
<div>
  text
  <level1>
    <level2>
      <level3>333</level3>
    </level2>
  </level1>
</div>;
  `,
  });
});

it("4 levels of nested jsx elements", () => {
  runTest({
    source: `
<div>
  text
  <level1>
    <level2>
      <level3>
        <level4>444</level4>
      </level3>
    </level2>
  </level1>
</div>;
  `,
  });
});

it("real world example", () => {
  runTest({
    source: `
<main>
  <div>
    <div>
      <h1>Sign in to GitHub</h1>
    </div>
    <div>
      <form>
        <label>Username or email address</label>
        <input name="login" />
        <div>
          <label>Password</label>
          <input name="password" />
          <input type="submit" name="commit" />
          <div>Don't remember password? <a>restore</a></div>
        </div>
      </form>
    </div>
    <p>
      New to GitHub?
      <a>Create an account</a>.
    </p>
  </div>
</main>
    `,
  });
});

it("jsx with params", () => {
  runTest({
    source: `
<span>Hi, {person.getName()}</span>
  `,
  });
});

it("import and require ignore", () => {
  runTest({
    source: `
import module1 from 'moduleName1';

const module2 = require('moduleName2');

function test1(a){
  function test2(b){
    function test3(c){
      return <div>text {a} {b} {c}</div>
    }
  }
}
  `,
  });
});
