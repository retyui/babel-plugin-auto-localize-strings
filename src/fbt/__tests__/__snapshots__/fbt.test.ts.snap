// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`2 levels of nested jsx elements 1`] = `
"import { fbt } from \\"fbt\\";
<div>
  <fbt desc=\\"description\\">
    text
    <level1>
      <level2>222</level2>
    </level1>
  </fbt>
</div>;
"
`;

exports[`3 levels of nested jsx elements 1`] = `
"import { fbt } from \\"fbt\\";
<div>
  <fbt desc=\\"description\\">
    text
    <level1>
      <level2>
        <level3>333</level3>
      </level2>
    </level1>
  </fbt>
</div>;
"
`;

exports[`4 levels of nested jsx elements 1`] = `
"import { fbt } from \\"fbt\\";
<div>
  <fbt desc=\\"description\\">
    text
    <level1>
      <level2>
        <level3>
          <level4>444</level4>
        </level3>
      </level2>
    </level1>
  </fbt>
</div>;
"
`;

exports[`add jsx wrapper on same deep level 1`] = `
"import { fbt } from \\"fbt\\";
<>
  <div>
    <div>
      <span>
        <fbt desc=\\"description\\">
          <a>
            <fbt desc=\\"description\\">1 link</fbt>
          </a>{\\" \\"}
          from friends 1
        </fbt>
      </span>
    </div>
  </div>
  <div>
    <div>
      <span>
        <fbt desc=\\"description\\">
          2 from friends <a>2 link</a>
        </fbt>
      </span>
    </div>
  </div>
</>;
"
`;

exports[`auto parametrization on end 1`] = `
"import { fbt } from \\"fbt\\";
<span>
  <fbt desc=\\"description\\">
    text <a>end</a>
  </fbt>
</span>;
"
`;

exports[`auto parametrization on start 1`] = `
"import { fbt } from \\"fbt\\";
<span>
  <fbt desc=\\"description\\">
    <a>
      <fbt desc=\\"description\\">start</fbt>
    </a>{\\" \\"}
    text
  </fbt>
</span>;
"
`;

exports[`ignore empty jsx text 1`] = `
"<div>
  <div />
</div>;
"
`;

exports[`import and require ignore 1`] = `
"import { fbt } from \\"fbt\\";
import module1 from \\"moduleName1\\";

const module2 = require(\\"moduleName2\\");

function test1(a) {
  function test2(b) {
    function test3(c) {
      return (
        <div>
          <fbt desc=\\"description\\">
            text {a} {b} {c}
          </fbt>
        </div>
      );
    }
  }
}
"
`;

exports[`jsx attributes 1`] = `
"import { fbt } from \\"fbt\\";
<input title={fbt(\\"need to translate\\", \\"description\\")} className=\\"ignore\\" />;
"
`;

exports[`jsx simple text 1`] = `
"import { fbt } from \\"fbt\\";
<span>
  <fbt desc=\\"description\\">Text</fbt>
</span>;
<span>{<fbt desc=\\"description\\">Text</fbt>}</span>;
<span>{<fbt desc=\\"description\\">Text</fbt>}</span>;
<span>{<fbt desc=\\"description\\">Text</fbt>}</span>;
"
`;

exports[`jsx with custom description 1`] = `
"import { fbt } from \\"fbt\\";
<span>
  <fbt desc=\\"my desc\\">Text</fbt>
</span>;
"
`;

exports[`jsx with params 1`] = `
"import { fbt } from \\"fbt\\";
<span>
  <fbt desc=\\"description\\">Hi, {person.getName()}</fbt>
</span>;
"
`;

exports[`literal variable 1`] = `
"import { fbt } from \\"fbt\\";
const test = fbt(\\"text\\", \\"description\\");
"
`;

exports[`literal variable if needed 1`] = `
"import { fbt } from \\"fbt\\";
const test = fbt(\\"text\\", \\"description\\");
const test2 = fbt(\\"text2\\", \\"my description\\");
"
`;

exports[`literal variable with custom description 1`] = `
"import { fbt } from \\"fbt\\";
const test = fbt(\\"text\\", \\"my desc\\");
"
`;

exports[`object property 1`] = `
"import { fbt } from \\"fbt\\";
const test = {
  prop: fbt(\\"test\\", \\"description\\"),
};
"
`;

exports[`prevent transform fbt options 1`] = `
"const test2 = fbt(\\"text2\\", \\"my description\\", {
  project: \\"foo\\",
});
"
`;

exports[`real world example 1`] = `
"import { fbt } from \\"fbt\\";
<main>
  <div>
    <div>
      <h1>
        <fbt desc=\\"description\\">Sign in to GitHub</fbt>
      </h1>
    </div>
    <div>
      <form>
        <label>
          <fbt desc=\\"description\\">Username or email address</fbt>
        </label>
        <input name={fbt(\\"login\\", \\"description\\")} />
        <div>
          <label>
            <fbt desc=\\"description\\">Password</fbt>
          </label>
          <input name={fbt(\\"password\\", \\"description\\")} />
          <input
            type={fbt(\\"submit\\", \\"description\\")}
            name={fbt(\\"commit\\", \\"description\\")}
          />
          <div>
            <fbt desc=\\"description\\">
              Don't remember password? <a>restore</a>
            </fbt>
          </div>
        </div>
      </form>
    </div>
    <p>
      <fbt desc=\\"description\\">
        New to GitHub?
        <a>Create an account</a>.
      </fbt>
    </p>
  </div>
</main>;
"
`;
